npep/npep-133.md

```diff
@@ -0,0 +1,89 @@
+# NPEP-133: FQDN Selector for Egress Traffic
+
+* Issue:
+  [#133](https://github.com/kubernetes-sigs/network-policy-api/issues/133)
+* Status: Provisional
+
+## TLDR
+
+This enhancement proposes adding a new selector to specify egress peers using
+Fully Qualified Domain Names (FQDNs). 
+
+## Goals
```

Comment:
Need to define the expected semantics as well; everyone _wants_ the semantics to be "all outbound connections are magically matched against the FQDN rules and it just works", but we know that won't really work.

- Do we support FQDN rules for both Allow and Deny rules?
  - If we support both Allow and Deny rules, what happens if, for example, someone allows connections to `example.com` but denies connections to `example.net` (which has the same IP).
- Do we require that pods make a DNS request before trying to connect, and do we require them to make that DNS request in a particular way (eg, using the resolver pointed to by the kubelet-provided `/etc/resolv.conf`)? Or do we expect the FQDN rule to work correctly even if the pod got the IP address via some means unknown to us?
  - (This is an example of the difference between Allow vs Deny rules: if we only support FQDN Allow rules, then we can say "the FQDN rule only works if you make a proper DNS request", and we can assume users will do that because they want their traffic to be allowed. But if we support FQDN Deny rules, we can't do that, because we have to assume that attackers will "cheat" in whatever way they can to get around the limits of the implementation.)
- Do we expect/require the rule to work correctly even when the domain name in question has rapidly-changing IPs and does not necessarily return all of its IPs in response to any single DNS query? (eg, if it returns different IPs to pods in one zone than it does to pods in another zone).

Reply:
> If we support both Allow and Deny rules, what happens if, for example, someone allows connections to example.com but denies connections to example.net (which has the same IP).

There is a pre-defined priority for such cases, except for one case: if the rules are listed in the same `to/from` list, which doesn't seem to have the priority semantics (yet?). But generally speaking, this problem also applies to the same pod being selected with 2 different labels, and having allow and deny rules for those. So I would say it is a valid, but probably separate problem.

> do we expect the FQDN rule to work correctly even if the pod got the IP address via some means unknown to us?

I am not sure that is possible, and would say no. As I see it, this feature is just a way to translate a dns name to a set of ips (would like to hear other opinions on this though). In that case, it makes sense to require a dns request to be sent for a dns name resolution. I imagine implementations may have more restrictions on how exactly this should happen, but I don't think we can come up with the common requirements.

> Do we expect/require the rule to work correctly even when the domain name in question has rapidly-changing IPs and does not necessarily return all of its IPs in response to any single DNS query? (eg, if it returns different IPs to pods in one zone than it does to pods in another zone).

Can this also be an implementation detail? I guess it doesn't change anything from the API point of view, and there are probably more corner-cases, e.g. replies with TTL=0 and so on.

Reply:
> > If we support both Allow and Deny rules, what happens if, for example, someone allows connections to example.com but denies connections to example.net (which has the same IP).
> 
> There is a pre-defined priority for such cases

No, priority only matters for rules that match the same things.

If I say "allow to 1.2.3.4 with priority 100" and "deny to 5.6.7.8 with priority 50" (and nothing else), then the priorities are irrelevant; all traffic to 1.2.3.4 is allowed, and all traffic to 5.6.7.8 is denied.

Likewise if I say "allow to redhat.com with priority 100" and "deny to google.com with priority 50"; all connections to redhat.com are allowed and all connections to google.com are denied.

So if I say "allow to example.com with priority 100" and "deny to example.net with priority 50", the expectation is that all traffic to example.com is allowed, and all traffic to example.net is denied, but what will _actually_ happen is that all traffic to both example.com and example.net will be denied, because the rules effectively become "allow to 93.184.216.34 with priority 100" and "deny to 93.184.216.34 with priority 50", and so the lower-precedence allow rule is overridden by the higher-precedence deny rule.

(There is really only one sane thing we can do here, which is to say "don't do that then", and a few of the implementations linked below are explicit about that. e.g., Azure notes that "FQDN filtering in application rules for HTTP/S ... can discern between two FQDNs that are resolved to the same IP address. This is not the case with FQDN filtering in network rules.")

> > do we expect the FQDN rule to work correctly even if the pod got the IP address via some means unknown to us?
> 
> I am not sure that is possible, and would say no.

In the case of not-rapidly-changing IPs, and non-wildcard, the policy implementation can resolve all the FQDNs to IPs ahead of time; this is how the existing OpenShift EgressFirewall feature works.

My point though was more that if we say "no" here, then that argues that we should _not_ support FQDN Deny rules, since I can't see any way to make them actually secure in that scenario.

> > Do we expect/require the rule to work correctly even when the domain name in question has rapidly-changing IPs
>
> Can this also be an implementation detail?

No, I don't think so; as I noted [below](https://github.com/kubernetes-sigs/network-policy-api/pull/134#discussion_r1331750773), both of the two implementations that use proactive/async DNS resolution also include warnings that they don't work well for rapidly-changing IPs, whereas none of the implementations that use DNS snooping have such a warning. So that implies that if we want to support the rapidly-changing IP use case, then are imposing certain technology choices on implementors.

Reply:
Well, OK, we _could_ leave "rapidly-changing IPs" as an implementation detail, and say that at the ANP API level, it's undefined whether they will work. But I would consider that very unsatisfying...

Reply:
>No, priority only matters for rules that match the same things.

Exactly, that is why we have priorities: to resolve "conflicts" for the same matches. If we could guarantee the same thing is never selected by different rules, we wouldn't need priorities at all. Your example kinda shows how priorities make these overlapping rules behave somewhat predictably. The same will happen if you "allow to pod selector label1=A with priority 100" and "deny to pod selector label2=B with priority 50", and then have a pod with labels `label1=A, label2=B`. So I think adding extra docs for dns examples, and for other similar examples will be useful here.

What I wanted to stress, is that there is a way to create rules saying `"allow to example.com with priority 100" and "deny to example.net with priority 100"` by just having these rules in the same ANP peers list (and the same can happen with the existing pod selectors). In that case we will have some kind of an undefined behaviour, because there is no strict prioritization between these rules, and the same thing will be allowed and denied with the same priority.

> Well, OK, we could leave "rapidly-changing IPs" as an implementation detail, and say that at the ANP API level, it's undefined whether they will work. But I would consider that very unsatisfying... 

It is certainly nice to have pre-defined behaviour, but it may be very difficult. For example, one plugin may support rapidly-changing ips at every 5 seconds rate, and another one at every millisecond rate. Defining such details will cause "imposing certain technology choices on implementors" as you said, which doesn't sounds like a flexible API.

Reply:
I don't know that "imposing certain technology choices on implementors" is a bad thing, so long as our decision is "reasonable" and backed by legitimate use-cases. This will be an optional spec of the API, so I hope it should be ok to consulting a bunch of providers and pick a compromise.

For example, the wild-card specification makes using a polling-based solution pretty much unworkable. But there is the fact that lots of providers have made it work, and  we see user-stories that ask for this -- both result in us considering it for the spec.

* In the rapidly-changing IP case, maybe it makes sense to specify a minimum TTL we expect providers to handle correctly (maybe 5 minutes). Providers can certainly do better if their implementation is capable.
* In the allow/deny case, I haven't heard of any user stories for deny that feel super rigorous to me. I'm of the opinion that deny shouldn't be part of the spec. If providers have deny implementations, I don't think keeping it out of the spec is a huge deal (users can still rely on the implementation-specific CRs if they really need it). Happy to reverse that if we find compelling use-cases.